sdk/servicebus/Azure.Messaging.ServiceBus/tests/Infrastructure/ServiceBusScope.cs

```diff
@@ -0,0 +1,627 @@
+            ///   Initializes a new instance of the <see cref="QueueScope"/> class.
+            /// </summary>
+            ///
+            /// <param name="serviceBusNamespaceName">The name of the Service Bus namespace to which the queue is associated.</param>
+            /// <param name="queueName">The name of the queue.</param>
+            /// <param name="shouldRemoveAtScopeCompletion">A flag indicating whether the queue should be removed when the scope is complete.</param>
+            ///
+            public QueueScope(string serviceBusNamespaceName,
+                              string queueName,
+                              bool shouldRemoveAtScopeCompletion)
+            {
+                NamespaceName = serviceBusNamespaceName;
+                QueueName = queueName;
+                ShouldRemoveAtScopeCompletion = shouldRemoveAtScopeCompletion;
+            }
+
+            /// <summary>
+            ///   Performs the tasks needed to remove the dynamically created Service Bus Queue.
+            /// </summary>
+            ///
+            public async ValueTask DisposeAsync()
+            {
+                if (!ShouldRemoveAtScopeCompletion)
+                {
+                    _disposed = true;
+                }
+
+                if (_disposed)
+                {
+                    return;
+                }
+
+                try
+                {
+                    var subscription = TestEnvironment.ServiceBusSubscription;
+                    var resourceGroup = TestEnvironment.ServiceBusResourceGroup;
+                    var token = await AquireManagementTokenAsync().ConfigureAwait(false);
+
+                    using var client = new ServiceBusManagementClient(new TokenCredentials(token)) { SubscriptionId = subscription };
+                    await CreateRetryPolicy().ExecuteAsync(() => client.Queues.DeleteAsync(resourceGroup, NamespaceName, QueueName)).ConfigureAwait(false);
+                }
+                catch
+                {
+                    // This should not be considered a critical failure that results in a test failure.  Due
+                    // to ARM being temperamental, some management operations may be rejected.  Throwing here
+                    // does not help to ensure resource cleanup only flags the test itself as a failure.
+                    //
+                    // If a queue fails to be deleted, removing of the associated namespace at the end of the
+                    // test run will also remove the orphan.
+                }
+
+                _disposed = true;
+            }
+        }
+
+        /// <summary>
+        ///   Provides access to dynamically created Service Bus Topic instance which exists only in the context
+        ///   of the scope; disposal removes the resource from Azure.
+        /// </summary>
+        ///
+        /// <seealso cref="IAsyncDisposable" />
+        ///
+        public sealed class TopicScope : IAsyncDisposable
```

Comment:
If we need the different types for queues vs topics, can we inherit from abstract class to avoid some of the property and method duplication.

Reply:
I don't see a benefit to using a hierarchy for something lightweight as this where only two parameters are really shared.   The only scenario that I can think of where you'd want to take a couple of scopes and treat them equally would be something like "I had a test with queues and topics and want to just dispose them all at once" which the interface supports - but is quite unlikely due to `using` syntax.